src/Nethermind/Nethermind.Blockchain/BlockTree.cs

```diff
@@ -771,6 +771,37 @@ private BlockHeader[] FindHeadersReversedFull(BlockHeader startHeader, int numbe
             return result;
         }
 
+        public BlockHeader FindLowestCommonAncestor(BlockHeader firstDescendant, BlockHeader secondDescendant)
+        {
+            if (firstDescendant.Number > secondDescendant.Number)
+            {
+                firstDescendant = GetAncestorAtNumber(firstDescendant, secondDescendant.Number);
+            }
+            else if (secondDescendant.Number > firstDescendant.Number)
+            {
+                secondDescendant = GetAncestorAtNumber(secondDescendant, firstDescendant.Number);
+            }
+
+            while (firstDescendant.Hash != secondDescendant.Hash)
+            {
+                firstDescendant = this.FindParentHeader(firstDescendant, BlockTreeLookupOptions.TotalDifficultyNotNeeded);
+                secondDescendant = this.FindParentHeader(secondDescendant, BlockTreeLookupOptions.TotalDifficultyNotNeeded);
+            }
+
+            return firstDescendant;
+        }
+
+        private BlockHeader GetAncestorAtNumber(BlockHeader header, long number)
+        {
```

Comment:
can stay like this - we may also think whether some optimization would be reasonable - if header is canonical wand we can ensure that during the lookup the UpdateMainChain is not called then we can simply load the canonical block agt number.